sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchange.scala

```diff
@@ -78,8 +82,15 @@ case class ShuffleExchange(
    * the returned ShuffleDependency will be the input of shuffle.
    */
   private[sql] def prepareShuffleDependency(): ShuffleDependency[Int, InternalRow, InternalRow] = {
-    ShuffleExchange.prepareShuffleDependency(
-      child.execute(), child.output, newPartitioning, serializer)
+    val dataSize = longMetric("dataSize")
+    val rdd = child.execute().mapPartitionsInternal { iter =>
+      val localDataSize = dataSize.localValue
+      iter.map { row =>
+        localDataSize.add(row.asInstanceOf[UnsafeRow].getSizeInBytes)
```

Comment:
Isn't this iteration over each row a significant added overhead? Seems it would be better to count the data size in bulk instead where the sort is done.